### ISSUE TITLE
Custom Unobtrusive validation not detected

### ISSUE THREAD

**Issue by jajvisser** [JeremySkinner/FluentValidation#1049]
### System Details

- FluentValidation version: 7.6.103
- Web Framework version  MVC 5 (5.2.3)

### Issue Description
We are trying to implement some custom unobtrusive validation, but we run into a weird issue with FluentValidation and custom PropertyValidators not being picked up. 

We have the following startup:
```
        FluentValidationModelValidatorProvider.Configure(provider =>
        {
            provider.ValidatorFactory = new UnityValidatorFactory();
            provider.Add(typeof(PostcodeValidator), (metadata, context, rule, validator) => new PostcodeClientPropertyValidator(metadata, context, rule, validator));
        });
```

My FluentValidationPropertyValidator:
```
public class PostcodeClientPropertyValidator : FluentValidationPropertyValidator
{
    public PostcodeClientPropertyValidator(ModelMetadata metadata, ControllerContext controllerContext, PropertyRule rule, IPropertyValidator validator) : base(metadata, controllerContext, rule, validator) { }

    public override IEnumerable<ModelClientValidationRule> GetClientValidationRules()
    {
        if (!ShouldGenerateClientSideRules())
            yield break;

        var formatter = new MessageFormatter().AppendPropertyName(Rule.PropertyName);
        var message = formatter.BuildMessage(Validator.ErrorMessageSource.GetString(null));

        yield return new ModelClientValidationRule
        {
            ErrorMessage = message,
            ValidationType = "validatePostcode",
        };
    }
}
```

My PropertyValidator
```
public class PostcodeValidator : PropertyValidator
{
    public PostcodeValidator(string errorMessage) : base(errorMessage) { }

    protected override bool IsValid(PropertyValidatorContext context)
    {
        var property = context.PropertyValue as string;
        return GeneralValidations.PostcodeRegex.IsMatch(property);
    }
}
```

My validator and extension
```
public static IRuleBuilderOptions<T, TProperty> MustValidateV21<T, TProperty>(this IRuleBuilder<T, TProperty> ruleBuilder, IValidationMessageTranslator translator)
{
    return ruleBuilder
         .SetValidator(new PostcodeValidator("Postcode invalid"))
         .WithMessage("Postcode invalid");
}
...
RuleFor(m => m.Postcode).MustValidatePostcode()
...
```

If i run this inside a normal Mvc Controller everything works as is should, but if i place it inside a sitecore controller the ModelValidator and PostcodeValidator still works, but the PostcodeClientPropertyValidator fails to run. The really weird part is that the default `FluentValidationPropertyValidators` are still working, but my own variants are not being picked up by FluentValidation. 

When i check at runtime whats in `ModelValidatorProviders.Providers` i see that all my `validatorFactories` have been registered correctly. 

Can anyone help me solve this issue?

**Comment by JeremySkinner** [comment]
Hi, it's hard to know without seeing more details. From the code you've posted everything looks ok, so I suspect there's something else happening in your application that's interfering with the process. If you put a breakpoint inside the `PostcodeClientPropertyValidator`s `GetClientValidationRules` method, is it ever invoked?

**Comment by jajvisser** [comment]
Hi Jeremy, No the code is not being invoked from a model within Sitecore. 
I also moved the delegate `(metadata, context, rule, validator) => new PostcodeClientPropertyValidator(metadata, context, rule, validator)` to a separate function to see if it was being called, but this also didn't happen.

**Comment by JeremySkinner** [comment]
Please can you put together a small sample project that reproduces the issue?

**Comment by jajvisser** [comment]
That a bit of an issue, i can't reproduce it outside of sitecore, in my simple sample projects it has been working fine. 

Can you perhaps show me where in the code i can look to see the correct code is being executed?

**Comment by jajvisser** [comment]
I think i just found my issue. It had todo with the way my viewmodel is constructed (and my validators and views):
`PersonsStepViewModel -> PersonViewModel (that implements a IAddress) -> Postcode`

In the same way i'm building up the view:
`PersonStepViewModel`
That has a `EditorFor` for `PersonViewModel` that has a `Partial` that contained `Postcode`.

So it looks like its related to a wierd bug you mentioned in this issue:
https://github.com/JeremySkinner/FluentValidation/issues/269

I will try to write a example application after this weekend to reproduce this bug.

**Comment by jajvisser** [comment]
It took a bit longer, but i made a small POC to replicate the bug i'm experiencing, i put it under my github.
https://github.com/jajvisser/poc-fluentvalidation-bug/tree/master

The /working variant works with fluentvalidation, but the /notworking variant has this bug.

**Comment by JeremySkinner** [comment]
Thanks for the sample. I see what's happening. 

Clientside validation does not pass through the same process as server-side validation when it comes to finding the correct validator to use. You're wiring up your child validator for the `MainPersonViewModel` by using ` .SetValidator(new MainPersonValidator())`. This works fine for serverside validation, but MVC generates clientside metadata for each model *in complete isolation* - MVC treats the `MainPersonModel` as a *root* model for the purposes of clientside validation, so the parent validator never exists (and therefore SetValidator is never called).

As you've already discovered in the "working" sample, if you annotate the `MainPersonModel` with a `Validator` attribute, then it works correctly because MVC is treating the `MainPersonViewModel` as a root-level model, and asks the validator provider to find the associated validator (which only exists if you've wired it up with the attribute). 

There is no way around this behaviour - it is just how MVC works. You'll need to pick one appraoch to use:
- Either annotate child models with the `[Validator]` attribute, which will allow for clientside validation inside partials like this
- Or use `SetValidator` and accept that clientside validation will not be available

You shouldn't mix the two approaches in an MVC5 project though, or you'll end up with double validation. If you choose to wire up child validators using the attribute then you should not use `SetValidator` in the parent.

**Comment by jajvisser** [comment]
Thank you for the explanation Jeremy, that makes total sense. 

In case your using dependency resolving to retrieve the validator, my guess is then to register the validator that way. Is that correct?

**Comment by JeremySkinner** [comment]
Yes that’s correct.

**Comment by jajvisser** [comment]
Hi Jeremy,

I have it completly figured out now for the 
Since ``IAddress`` is a interface you can't retrieve the Validator by the interface, so you need to register the interface to the Validator (which is not possible with ``[Validator]``). 

After a bit of digging, i found that ``FluentValidationModelValidatorProvider`` creates the validator 
In my case that is not completly possible, i need to be able to find the correct validator for ``IAddress`` since i'm using it over multiple models. 

```
protected virtual IValidator CreateValidator(ModelMetadata metadata, ControllerContext context) {
	if (IsValidatingProperty(metadata)) {
		return ValidatorFactory.GetValidator(metadata.ContainerType);
	}
	return ValidatorFactory.GetValidator(metadata.ModelType);
}
```
Looking at this code, it should be possible to see the exact model the ``ValidatorFactory`` is trying to find, but is the ``metaData`` is not available in the ``ValidatorFactory``. 

This dirty hack solved my problem for now:
```
public class FluentValidationModelMetaDataValidatorProvider : FluentValidationModelValidatorProvider
{
	public new static void Configure(Action<FluentValidationModelValidatorProvider> configurationExpression = null)
	{
		configurationExpression = configurationExpression ?? delegate { };

		var provider = new FluentValidationModelMetaDataValidatorProvider();
		configurationExpression(provider);

		DataAnnotationsModelValidatorProvider.AddImplicitRequiredAttributeForValueTypes = false;
		ModelValidatorProviders.Providers.Add(provider);
	}

	protected override IValidator CreateValidator(ModelMetadata metadata, ControllerContext context)
	{
		if (ValidatorFactory is UnityValidatorFactory)
		{
			((UnityValidatorFactory) ValidatorFactory).Metadata = metadata;
			((UnityValidatorFactory) ValidatorFactory).Context = context;
		}

		if (IsValidatingProperty(metadata))
		{
			return ValidatorFactory.GetValidator(metadata.ContainerType);
		}
		return ValidatorFactory.GetValidator(metadata.ModelType);
	}
}
```

Initialization:
```
FluentValidationModelMetaDataValidatorProvider.Configure(provider =>
{
	provider.ValidatorFactory = new UnityValidatorFactory();
        ...
});
```

UnityValidatorFactory:
```
public class UnityValidatorFactory : ValidatorFactoryBase
{
	public ModelMetadata Metadata { get; set; }
	public ControllerContext Context { get; set; }

	public override IValidator CreateInstance(Type validatorType)
	{
		try
		{
			if (validatorType.GenericTypeArguments[0].IsInterface)
			{
				validatorType = typeof(IValidator<>).MakeGenericType(Metadata.Container.GetType());
			}
			return DependencyResolver.Current.GetService(validatorType) as IValidator;
		}
		catch (Exception)
		{
			return null;
		}
	}
}
```

My suggestion is to change the method signature of ``ValidatorFactory.GetValidator(Type validatorType)`` to ``ValidatorFactory.GetValidator(Type validatorType, ModelMetadata metadata)``

**Comment by JeremySkinner** [comment]
Changing be GetValidator signature isn’t feasible or desirable here. The interface for IValidatorFactory is in the core library which intentionally has no reference to MVC and it’s types. The interface represents the concept of “give me a validator for a specific type”. 

So basically: move your interface check and generic type construction to the provider and return the factory to its previous state. 

Overriding the ModelValidatorProvider is the correct approach, but you should be putting the customisation in there not in the ValidatorFactory. The factory should simply be responsible for creating a validator for the type it’s given. The selection logic for the type should be in your ModelValidatorProvider. Also, setting properties on the factory like that is extremely dangerous as this isn’t thread safe.

**Comment by jajvisser** [comment]
Thank you Jeremy, i'm glad we solved this issue. The seperation between web and core shouldn't be crossed. This comment above was just for a proof of concept. 

I will add my final solution later, for possible future reference. Then i will close this issue.

**Comment by jajvisser** [comment]
Here is my solution, i added a interception to ``CreateValidator`` change the modeltype in the ``CustomFluentValidationModelValidatorProvider``. 

```
public class CustomFluentValidationModelValidatorProvider : FluentValidationModelValidatorProvider
{
	public Func<Type, bool, ModelMetadata, ControllerContext, Type> CreateValidatorTypeInterceptor { get; set; }

	public static void Configure(Action<CustomFluentValidationModelValidatorProvider> configurationExpression = null)
	{
		configurationExpression = configurationExpression ?? delegate { };

		var provider = new CustomFluentValidationModelValidatorProvider();
		configurationExpression(provider);

		DataAnnotationsModelValidatorProvider.AddImplicitRequiredAttributeForValueTypes = false;
		ModelValidatorProviders.Providers.Add(provider);
	}

	/// <summary>
	/// This is a workaround to intercept the ModelValidatorProvider.CreateValidator parameters. 
	/// Now you can add a CreateValidatorTypeInterceptor in the configure method to intercept the CreateValidator and change the model type with custom logic.
	/// </summary>
	/// <param name="metadata">container of the current model and property metadata</param>
	/// <param name="context">controllercontext of the current view</param>
	/// <returns></returns>
	protected override IValidator CreateValidator(ModelMetadata metadata, ControllerContext context)
	{
		var validatorType = metadata.ModelType;
		var isProperty = false;
		if (IsValidatingProperty(metadata))
		{
			validatorType = metadata.ContainerType;
			isProperty = true;
		}

		if (CreateValidatorTypeInterceptor != null)
		{
			validatorType = CreateValidatorTypeInterceptor(validatorType, isProperty, metadata, context);
		}

		return ValidatorFactory.GetValidator(validatorType);
	}
}
```

Initalization:
```
CustomFluentValidationModelValidatorProvider.Configure(provider =>
{
	provider.ValidatorFactory = new UnityValidatorFactory();
	provider.CreateValidatorTypeInterceptor = (validatorType, isProperty, metadata, context) =>
	{
		if (isProperty && validatorType.IsInterface)
		{
			validatorType = metadata.Container.GetType();
		}

		return validatorType;
	};
        ...
});
```

This solved my issue, so i'm closing this issue.
